src/nitllvm.nit

```diff
@@ -0,0 +1,155 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# LLVM backend for the Nit compiler (Work in progress)
+#
+# # Features and TODO list
+#
+# - [x] Generate basic class structures
+# - [x] Generate LLVM function signatures
+# - [ ] Generate LLVM implementation code
+# - [ ] Attributes
+# - [ ] Intern methods
+# - [ ] Generic and other types
+# - [ ] Support the FFI with C
+#
+# # References
+#
+# * http://llvm.org/docs/WritingAnLLVMBackend.html
+# * http://llvm.org/docs/LangRef.html
+# * http://llvm.org/releases/2.6/docs/tutorial/JITTutorial1.html
+module nitllvm
+
+import modelbuilder
+import frontend
+import model_utils
+
+redef class ToolContext
+	# --output
+	var opt_output = new OptionString("Output file", "-o", "--output")
+
+	init do option_context.add_option opt_output
+
+	# LLVM code generation
+	var llvm_phase: Phase = new LLVMPhase(self, null)
+end
+
+private class LLVMPhase
+	super Phase
+
+	redef fun process_mainmodule(mainmodule, mmodules)
+	do
+		# Prepare LLVM code in memory
+		var llvm_classes_vfts = new HashMap[String, String]
+		var llvm_functions = new HashMap[String, String]
+		var llvm_attributes = new HashMap[String, String]
+
+		for mclass in mainmodule.flatten_mclass_hierarchy do
+			var class_vft = ["%class_{mclass.name}_vft = type \{\n"]
+			var class_vft_entries = new Array[String]
+			var class_attributes = ["\n%class_{mclass.name} = type \{%class_{mclass.name}_vft*"] # TODO
+
+			var mproperties = mclass.all_mproperties(mainmodule, private_visibility)
+			for mproperty in mproperties do
+				if mproperty isa MMethod then
+
+					var mpropdefs = mproperty.lookup_all_definitions(mainmodule, mclass.mclass_type)
+					for mpropdef in mpropdefs do
+
+						var class_vft_entry = ["\t"]
+
+						var index = "{mpropdef.mclassdef.name}_{mpropdef.name}"
+						var llvm_function = ["define "]
+						if mpropdef.msignature.return_mtype != null then
+							llvm_function.add "%{mpropdef.msignature.return_mtype.name} "
+							class_vft_entry.add "%{mpropdef.msignature.return_mtype.name}("
+						else
+							llvm_function.add "void "
+							class_vft_entry.add "void("
+						end
+
+						var params = new Array[String]
+
+						# Receiver
+						params.add "%class_{mpropdef.mclassdef.name}*"
+						llvm_function.add "@{index}(%class_{mclass.name}* %this"
+
+						for param in mpropdef.msignature.mparameters do
+							llvm_function.add ",%class_{param.mtype.name} %{param.name}"
+							params.add "%class_{param.mtype.name}"
+						end
+						class_vft_entry.add params.join(",")
+
+						llvm_function.add """) nounwind {
```

Comment:
@privat this attribut (or annotation) is used to indicate that the function never raises an exception !